File: Project_Java/src/Diagram.java
import java.util.ArrayList;
import java.util.List;

public class Diagram {
    List<State> T;
    List<Edge> E;

    Diagram(Grammar grammar){
        //Initialize the graph
        T = new ArrayList<>();
        E = new ArrayList<>();

        //Add the first state
        State root = new State(0);
        root.rules.add(new PositionRule(grammar.rules.get(0),0));
        root.closure(grammar);
        T.add(root);

        int i=0;
        while (i < T.size()) {
            State I = T.get(i);
            for(PositionRule rule:I.rules){
                if(!rule.isLast()) {
                    String X = rule.getNextSymbol();
                    if(!X.equals("$")) {
                        State J = Goto(I, X, grammar);
                        //Check if there is already a state with the same rules
                        boolean repeated = false;
                        for(State s:T){
                            if (equalStates(s,J)) {
                                boolean edgePresent= false;
                                for(Edge e:E){
                                    if (e.start == I.id && e.end == s.id) {
                                        edgePresent = true;
                                        break;
                                    }
                                }
                                if(!edgePresent) {
                                    E.add(new Edge(X, I.id, s.id));
                                }
                                repeated = true;
                                break;
                            }
                        }
                        if(!repeated) {
                            T.add(J);
                            E.add(new Edge(X, I.id, J.id));
                        }
                    }
                }
            }
            i++;
        }
    }

    private boolean equalStates(State s, State J) {
        if (s.rules.size() != J.rules.size()){
            return false;
        }
        for(int i=0; i< s.rules.size(); i++) {
            PositionRule pr1 = s.rules.get(i);
            PositionRule pr2 = J.rules.get(i);
            if (pr1.position != pr2.position || !pr1.rule.equals(pr2.rule) ) {
                return false;
            }
        }
        return true;
    }

    public State Goto(State I, String X, Grammar grammar) {
        State J = new State(T.size());
        for (PositionRule rule:I.rules) {
            if(!rule.isLast() && rule.getNextSymbol().equals(X)){
                J.rules.add(new PositionRule(rule.rule, rule.position+1));
            }
        }
        J.closure(grammar);
        return J;
    }
}

File: Project_Python/src/Diagram.py
from Edge import Edge
from PositionRule import PositionRule
from State import State


def equalStates(s, J):
    if not len(s.rules) == len(J.rules):
        return False
    for i in range(len(s.rules)):
        pr1 = s.rules[i]
        pr2 = J.rules[i]
        if (not pr1.position == pr2.position) or (not pr1.rule == pr2.rule):
            return False
    return True


class Diagram:
    def __init__(self, grammar):
        # Initialize the graph
        self.T = []
        self.E = []

        # Add the first state
        root = State(0)
        root.rules.append(PositionRule(grammar.rules[0], 0))
        root.closure(grammar)
        self.T.append(root)

        i = 0
        while i < len(self.T):
            IState = self.T[i]
            for rule in IState.rules:
                if not rule.isLast():
                    X = rule.getNextSymbol()
                    if not X == "$":
                        J = self.Goto(IState, X, grammar)
                        # Check if there is already a state with the same rules
                        repeated = False
                        for s in self.T:
                            if equalStates(s, J):
                                edgePresent = False
                                for e in self.E:
                                    if e.start == IState.id and e.end == s.id:
                                        edgePresent = True

                                if not edgePresent:
                                    self.E.append(Edge(X, IState.id, s.id))

                                repeated = True
                                break

                        if not repeated:
                            self.T.append(J)
                            self.E.append(Edge(X, IState.id, J.id))
            i += 1

    def Goto(self, IState, X, grammar):
        J = State(len(self.T))
        for rule in IState.rules:
            if (not rule.isLast()) and rule.getNextSymbol() == X:
                J.rules.append(PositionRule(rule.rule, rule.position + 1))
        J.closure(grammar)
        return J
